File: rustnlpjava/src/main/java/dev/gigapixel/rustnlp/Tok.java
package dev.gigapixel.rustnlp;

import dev.gigapixel.rustnlp.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Tok {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class TokCleaner implements Runnable {

        MemorySegment segment;
        TokCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            rustnlp_h.Tok_destroy(this.segment);
        }
    }

    Tok() {}
    Tok(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    public static Tok create() {
        
        var nativeVal = rustnlp_h.Tok_create();
        
        List<Object> selfEdges = List.of();
        
        
        
        var returnVal = new Tok(nativeVal, selfEdges);
        return returnVal;
                
    }
    
    
    public TokRes tokenize(String text) {
        
        try (var arena = Arena.ofConfined()) {
            
            var textData= Arena.ofAuto().allocateFrom(text, StandardCharsets.UTF_8);
            var textLen = textData.byteSize() - 1;  // allocated strings are null terminated
            var textView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(textView, textLen);
            DiplomatStringView.data(textView, textData);
            var nativeVal = rustnlp_h.Tok_tokenize(internal, textView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new TokRes(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public EncodeResult tokenizeAsOffsets(String text) {
        
        try (var arena = Arena.ofConfined()) {
            
            var textData= Arena.ofAuto().allocateFrom(text, StandardCharsets.UTF_8);
            var textLen = textData.byteSize() - 1;  // allocated strings are null terminated
            var textView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(textView, textLen);
            DiplomatStringView.data(textView, textData);
            var nativeVal = rustnlp_h.Tok_tokenize_as_offsets(internal, textView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new EncodeResult(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
}

File: rustnlpkotlin/src/main/kotlin/dev/gigapixel/rustnlp/Tok.kt
package dev.gigapixel.rustnlp;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure


internal interface TokLib: Library {
    fun Tok_destroy(handle: Pointer)
    fun Tok_create(): Pointer
    fun Tok_tokenize(handle: Pointer, text: Slice): Pointer
    fun Tok_tokenize_as_offsets(handle: Pointer, text: Slice): Pointer
}

class Tok internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>
)  {

    internal class TokCleaner(val handle: Pointer, val lib: TokLib) : Runnable {
        override fun run() {
            lib.Tok_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<TokLib> = TokLib::class.java
        internal val lib: TokLib = Native.load("rustnlp", libClass)
        
        fun create(): Tok {
            
            val returnVal = lib.Tok_create();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Tok(handle, selfEdges)
            CLEANER.register(returnOpaque, Tok.TokCleaner(handle, Tok.lib));
            return returnOpaque
        }
    }
    
    fun tokenize(text: String): TokRes {
        val (textMem, textSlice) = PrimitiveArrayTools.readUtf8(text)
        
        val returnVal = lib.Tok_tokenize(handle, textSlice);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = TokRes(handle, selfEdges)
        CLEANER.register(returnOpaque, TokRes.TokResCleaner(handle, TokRes.lib));
        textMem.close()
        return returnOpaque
    }
    
    fun tokenizeAsOffsets(text: String): EncodeResult {
        val (textMem, textSlice) = PrimitiveArrayTools.readUtf8(text)
        
        val returnVal = lib.Tok_tokenize_as_offsets(handle, textSlice);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = EncodeResult(handle, selfEdges)
        CLEANER.register(returnOpaque, EncodeResult.EncodeResultCleaner(handle, EncodeResult.lib));
        textMem.close()
        return returnOpaque
    }

}
